src/codegen/machine_ir.h

```diff
@@ -0,0 +1,180 @@
+  ALL_IR_INSTRUCTION_TYPES(DEFINE_MIR_INSTRUCTION_TYPE)
+  /// Marks beginning of block
+  // TODO: Do we need this?
+  MIR_BLOCK,
+  MIR_COUNT,
+  MIR_ARCH_START = 0x420
+} MIROpcodeCommon;
+#undef DEFINE_MIR_INSTRUCTION_TYPE
+
+typedef enum MIROperandKind {
+  MIR_OP_NONE,
+  MIR_OP_REGISTER,
+  MIR_OP_IMMEDIATE,
+  MIR_OP_BLOCK,
+  MIR_OP_FUNCTION,
+  MIR_OP_NAME,
+  MIR_OP_COUNT
+} MIROperandKind;
+
+typedef struct MIROperandRegister {
+  uint32_t value;
+  uint16_t size;
+} MIROperandRegister;
+typedef const char* MIROperandName;
+typedef int64_t MIROperandImmediate;
+typedef IRBlock* MIROperandBlock;
+typedef IRFunction* MIROperandFunction;
+
+typedef unsigned int MIRRegister;
+
+typedef struct MIRValue_x86_64 {
+  uint8_t instruction_form;
+  uint16_t instruction;
+  uint8_t reg_src_sz;
+  uint8_t reg_src;
+  uint8_t reg_dst_sz;
+  uint8_t reg_dst;
+  uint8_t reg_addr;
+  int64_t immediate;
+  int64_t offset;
+  const char *name;
+  IRBlock *ir_block;
+  IRFunction *ir_function;
+} MIRValue_x86_64;
+
+typedef struct MIROperand {
+  MIROperandKind kind;
+  union {
+    MIROperandRegister reg;
+    MIROperandImmediate imm;
+    MIROperandName name;
+    MIROperandBlock block;
+    MIROperandFunction function;
+  } value;
+} MIROperand;
+
+#define MIR_OPERAND_SSO_THRESHOLD 3
+
+// NOTE: There is no way of keeping track of which ISA a particular
+// instruction is for; that's up to each backend that uses it!
+typedef struct MIRInstruction {
+  size_t id;
+
+  MIRRegister reg;
+
+  /// An MIR opcode is some integer that has the 1:1 IR types from 0->N
+  /// and then after that each ISA starts it's own opcode enum for
+  /// architecture-specific instructions.
+  /// The idea is that each ISA gets fed an MIR full of regular IR
+  /// instructions and it's job is to translate them into (or build a
+  /// new MIR with) the arch-specific opcodes.
+  uint32_t opcode;
+
+  uint8_t operand_count;
+  union {
+    MIROperand arr[MIR_OPERAND_SSO_THRESHOLD];
+    Vector(MIROperand) vec;
+  } operands;
+
+  // Architecture-specific values.
+  union {
+    MIRValue_x86_64 x64;
+  };
+
+  MIRBlock *block;
+
+  // Keep track of originating IR instruction.
+  IRInstruction *origin;
+
+  // If an architecture lowers a generically-lowered MIR instruction,
+  // store the arch-lowered MIR instruction in the generic instruction
+  // so that we may properly update references going forward.
+  struct MIRInstruction *lowered;
+
+} MIRInstruction;
+
+typedef struct MIRBlock {
+  string name;
+
+  // Index within `instructions` of MIRFunction pointing to first
+  // instruction executed in basic block.
+  MIRRegister entry;
+  // Index within `instructions` of MIRFunction pointing to last
+  // instruction of basic block.
+  MIRRegister exit;
+
+  MIRFunction *function;
+
+  IRBlock *origin;
+} MIRBlock;
+
+typedef struct MIRFunction {
+  string name;
+
+  /// Flat list of all instructions in function.
+  MIRInstructionVector instructions;
```

Comment:
This being a vector rather than a list definitely makes sense, and it might well make sense to perform instruction selection into a completely new vector because we only ever have to modify the last couple instructions during ISel. We may have to split this into several vectors though, one for each block (see also the previous comment in this file)

Reply:
How then would one find an instruction by virtual register?

Right now we use the virtual register of an instruction as an index into this vector (and we use the size of this vector to set the virtual register when pushing a new instruction). This means that all the virtual register references have constant lookup and (eventual) MIR register allocation would have a much quicker time resolving all these references. Why do the instructions need to be stored in separate vectors?

Reply:
> How then would one find an instruction by virtual register?

Not at all. Assuming the RA works properly, there is no need for that since instructions communicate over what registers they use and place their results in

Reply:
> Why do the instructions need to be stored in separate vectors?

Mostly for the reasons discussed in the previous comment in this file (if we change how blocks are handled)

Reply:
> > How then would one find an instruction by virtual register?
> 
> Not at all. Assuming the RA works properly, there is no need for that since instructions communicate over what registers they use and place their results in

Interesting. So you're saying RA would keep a "translation list" of vreg->hwreg and then any time it encounters that vitrual register it just looks it up in that list and replaces it with the corresponding hardware register? And in this way a virtual register operand would be resolved without having to reference the machine instruction that populates that register at all

Reply:
The RA’s job is to assign hw regs to the vregs, but yes